### DISCUSSION TITLE
Also trying to get running on .net Maui...

### DISCUSSION THREAD

**Discussion by submarines414** [TolikPylypchuk/SharpHook#22] [Q&A] [answered]
This is just for desktop applications though. I am completely oblivious/new to async programming patterns. I understand starting my own task and awaiting a result just fine, but awaiting a keyboard listener that should run for the life of the program (and therefore there is nothing to await) just confuses me more. After using RunAsync() on an initialized TaskPoolGlobalHook with an applicable event listener:

void OnKeyPressed(object sender, KeyboardHookEventArgs e) { // assign string to some GUI field here }

I get a peculiar exception message:

Exception thrown: 'System.Runtime.InteropServices.COMException' in WinRT.Runtime.dll
An exception of type 'System.Runtime.InteropServices.COMException' occurred in WinRT.Runtime.dll but was not handled in user code
The application called an interface that was marshalled for a different thread. (0x8001010E (RPC_E_WRONG_THREAD))

I have done a similar callback system in a basic console test app. I still had an error when I tried to call the destructor with one of the keys being listened itself (something about a referenced item not existing), but the key codes were able to be printed to the console just fine.

You mentioned the singleton nature of hardware listeners in some of your docs and I'm wondering if maybe it's conflicting with the MAUI system input listeners themselves, but probably I just have no idea how to use an async function. Obviously I can't use Run() the regular way because that will just block my entire app from updating, unless I run and destroy it for every keypress. 

Are there any current examples of using the non-react version in an application like this?

Background: used to being spoonfed key codes from useful Rust libraries and game engines and am a newb with c#.

**Comment by TolikPylypchuk** [comment, marked as answer]
Hi! Thanks for using SharpHook!

It looks like a problem with threads. You see, when you call `RunAsync` on a `TaskPoolGlobalHook` (or `SimpleGlobalHook`) it will start a new thread and run the global hook there. In case of `SimpleGlobalHook`, all event handlers will also run in that same new thread. In case of `TaskPoolGlobalHook`, the event handlers will run on some thread in the thread pool for `Task` objects. Sounds difficult to someone who's new to this, but you don't have to dive into details. All you have to know is that if you want to update the UI in MAUI (or any other UI framework for .NET) you'll have to do that in the main thread. You can use the [`MainThread`](https://learn.microsoft.com/en-us/dotnet/maui/platform-integration/appmodel/main-thread) class for this.

Your code will then look something like this:

```c#
void OnKeyPressed(object sender, KeyboardHookEventArgs e)
{
    MainThread.BeginInvokeOnMainThread(() =>
    {
        // assign string to some GUI field here
    });
}
```

If you have any problems, feel free to continue the discussion.

**Reply by submarines414** [reply]
Oh that makes sense! Thank you so much, I super appreciate it, as well as this code you've written.

**Comment by submarines414** [comment]
Okay I've been running this successfully on windows .net 6 and 7 for a while, but the same code on mac (visual studio, etc) seems to make the hook and then the thread immediately drops. Have you seen anything similar from others? It's the same environment as before, using with .net maui apps. Using the non-react, happens with both task pool and simple, already made sure no parts of the program are accidentally disposing it while the program's running.

**Reply by submarines414** [reply]
9  0x0 in System.SR.get_Arg_DllNotFoundException

When I debugged and looked at threads, the one it was running on threw this. When I invoked Run() for the TaskPoolGlobalHook on the main thread, I got a little window but it didn't have much info. Says something about "reflection"...

The calls it cycles through (I haven't been able to figure out how to debug threads too well) is:
-> System.Threading.ExecutionContext.get_IsDefault
-> System.Threading.Thread.StartHelper.RunWorker
-> System.Threading.Thread.StartHelper.InitializeCulture
-> System.Threading.AutoReleasePool.get_EnableAutoReleasePool
-> SharpHook.Native.UioHook.SetDispatchProc
-> System.SR.get_Arg_DllNotFoundException

**Reply by submarines414** [reply]
Okay more stuff: was able to get this out of a exception window:
	
InnerException	{System.DllNotFoundException: uiohook    at SharpHook.GlobalHookBase.Run()}	System.DllNotFoundException

the dll not found is uiohook, which is from the base C lib this project wraps, right?

**Reply by TolikPylypchuk** [reply]
Hi! Hm, that's weird. Could you please check the content of the output directory? Usually it's `bin/Debug/net7.0`. There should be a file named `libuiohook.dylib` - this is the libuiohook binary for macOS. If it's not there, then it means that the package didn't extract it into the output directory correctly. If it is present, then it's really weird and we'll have to investigate further. Problem is, I don't have a Mac, so it will be difficult for me to figure this out. As a workaround, you can copy this file from the package into the output directory manually.

One other thing - the process that uses SharpHook on macOS must have accessibility API enabled - by default the OS won't let you create a hook as a security measure. If you call `hook.Run()` or `hook.RunAsync()` it will throw `HookException` with the value `AxApiDisabled` and a system popup will appear which will tell you that the process tried to use the accessibility API. You can enable it in the OS settings. This is just FYI, so that you know it's expected behavior.

**Reply by submarines414** [reply]
Hmmm... the system popup never occurred, the exception type was definitely HookException but I don't recall seeing AxApiDisabled flag anywhere. 

After having checked for libuiohook.dylib... it is not present in the debug bin folder. Is there a project location it should have extracted to outside of the build path when I installed the nuget package? I'll try copying it from your project to that folder and see what happens.

**Reply by submarines414** [reply]
Maybe a dumb question, but the compiled binary for mac -- libuiohook.dylib -- I'm not able to find that compiled anywhere. When I download your project, the symbolic folder link to libuiohook C proj downloads as an empty dir. Should I be following that link and trying to compile it with make?

**Reply by TolikPylypchuk** [reply]
Regarding the system popup and `AxApiDisabled` - forget about it for the moment, it will come up after we fix the first issue :)

OK, so `libuiohook.dylib` is not present in the output folder - that the real problem and one I'll have to fix. The file itself is contained in the NuGet package; it's just not copied to where it's needed. You can find the contents of the package in the hidden folder `.nuget` in your home directory - it's actually the folder into which NuGet downloads all packages. Inside `.nuget` go to `packages/sharphook/4.0.0/runtimes/osx-<platform>/native` (where `<platform>` is `x64` if you have an Intel CPU, or `arm64` if you have an Apple CPU, like M1 or M2). `libuiohook.dylib` will be there. Copy it into `bin/Debug/net7.0` of your project. As I've said, it should be done automatically, but for some reason it isn't.

**Reply by submarines414** [reply]
got it -- thanks so much for the help!

**Reply by TolikPylypchuk** [reply]
Tried to reproduce the issue and realized that you're getting this error on macOS. I don't have a Mac, so I can't currently reproduce this. Could you please create an issue about it, so I don't forget to look into it if I get an access to a Mac?

**Reply by karthitect** [reply]
So I'm running into this as well on macOS Ventura (m1). I tried physically copying libuiohook.dylib to bin/Debug/net7.0-maccatalyst/maccatalyst-x64/ but my hooks are not firing. I don't see any exceptions either. This setup works fine on Windows 11.

Any suggestions? Thanks!

**Reply by TolikPylypchuk** [reply]
Hm, I'm not sure libuiohook (and SharpHook) works on Mac Catalyst. Could you try to run the macOS build of the app instead and see whether it works?

**Reply by karthitect** [reply]
I'm using .NET MAUI for cross platform desktop development and it appears that MAUI only supports maccatalyst. I'll do some troubleshooting when I get chance.

Thanks for the quick response..

**Reply by karthitect** [reply]
Ok, now I'm getting this exception:

{System.DllNotFoundException: uiohook
   at SharpHook.Native.UioHook.SetDispatchProc(DispatchProc dispatchProc, IntPtr userData)
   at SharpHook.GlobalHookBase.<>c__DisplayClass16_0.<RunAsync>b__0()}
   
I've double checked and triple checked that I have libuiohook.dylib in bin/Debug/net7.0-maccatalyst/maccatalyst-x64/. And also made sure that I copied the osx arm64 version. Not sure why it's unable to load the native library.

Here's the StackTrace

at [redacted].Maui.App.Window_Created(Object sender, EventArgs e) in App.xaml.cs:line 98\n
at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__128_0(Object state)\n
at Foundation.NSAsyncSynchronizationContextDispatcher.Apply()\n--- End of stack trace from previous location ---\n
at UIKit.UIApplication.UIApplicationMain(Int32 argc, String[] argv, IntPtr principalClassName, IntPtr delegateClassName)\n  
at UIKit.UIApplication.Main(String[] args, Type principalClass, Type delegateClass)\n
at Program.Main(String[] args) in [redacted]

**Reply by TolikPylypchuk** [reply]
It appears that libuiohook needs to be built for Mac Catalyst as well. I'm not sure how to achieve that and whether its functionality will work at all on Mac Catalyst. I also don't have access to a Mac at the moment, so I won't be able to test any improvements. I'll create an issue so that it's more visible. Let's move the discussion there.

**Reply by headintheclouds21** [reply]
I've also been trying to use this for Mac Catalyst and I've been running into the same issue as @karthitect . I've placed the dlls in bin/Debug/net7.0-maccatalyst/maccatalyst-x64 and they're still not being found. Does anyone have any advice for this? Has there been any progress on this? I'd love to use SharpHook in my NET MAUI application, but this is a large impediment that's preventing me from doing so.

**Reply by TolikPylypchuk** [reply]
@headintheclouds21 Thanks for asking, but unfortunately no, there's been no progress on SharpHook support for Mac Catalyst. This is being tracked in issue #35.

**Reply by TolikPylypchuk** [reply]
FYI, I've just released version 5.0.0 which adds support for .NET MAUI on macOS.

**Reply by headintheclouds21** [reply]
@TolikPylypchuk Awesome! That's exciting! When I get a chance I'm going to try using it
